Feature: vendor push

sharnik opened the pull request:
Here's implementation for `vendor push`. It currently just does git push --all and git push --tags. 
This will change when we add notes with repository state information.

Also, it adds multi_json as runtime dependency, as it was missing.

sharnik committed:
Add multi_json as runtime dependency.

sharnik committed:
Add `vendor push` command.

sharnik committed:
Cucumber feature for `vendor push`.

sharnik commented:
Travis says there's a timeout on `vendor sync` on RBX. Maybe we should extend the timeout threshold, because it seems to fail often? I've done that for MRI in the other PR. What do you think?

mpasternacki committed:
Change require to autoload for multi_json
(cherry picked from commit bc2af1c38bf32f5fb5c8448b437f674a30fc43cb)

mpasternacki commented:
+1 for increasing timeout. Maybe we should just bump it to something ridiculous like 1 minute on all implementations?

Re multi_json dependency: I've changed `require` to `autoload` on develop. This is needed only for `chef_cookbook`, which needs to load Chef, which depends on MultiJson, and is deliberately left out of gem dependencies to avoid blowing them up. Chef cookbook vendor will likely end up in a separate gem anyway to clean up dependencies.

`vendor push` should call `Environment#each_vendor_instance` and push only branches managed by Vendorificator. It can be done in a single git call, together with `--tags`, by specifying a list of refspecs. Just using `git push --all` assumes that user doesn't have any feature branches they prefer to keep local. We'll need to iterate over modules/branches anyway to push notes.

sharnik committed:
Bump timeouts to ridiculous amounts for Travis.

sharnik committed:
Revert "Add multi_json as runtime dependency."

This reverts commit ef39aa8e28e6f7bd6b72beb0f333e3e088881a37.

sharnik commented:
Timeouts got ricidulous and I've reverted the multi_json dependency. More to come.

sharnik committed:
Do a per-module push rather than global one.

sharnik committed:
Merge remote-tracking branch 'origin/develop' into feature/push-command

sharnik committed:
Freeze aruba at old (working) version.

Would be good to review the changes introduced in 0.5.2. Maybe
there's something useful for us.

sharnik commented:
Finally Travis's green. Ready to merge.

mpasternacki commented:
Issuing one git command per module is not necessary. It can be done more efficiently in one command for both branches and tags by collecting modules and specifying a list of refspecs for each module's branch and all of its tags (I've included `refs/heads` / `refs/tags` prefix in the example to avoid ambiguity - this is plumbing, it's better to be ugly but precise than rely on user-friendly shortcuts):

```
git push origin +refs/heads/vendor/foo +refs/tags/vendor/foo/1 +refs/tags/vendor/foo/2 +refs/heads/vendor/bar +tags/vendor/bar/23.1
```

sharnik committed:
One command to push them all.

sharnik commented:
You were absolutely right, I forgot that there'll probably be shitloads of modules. Fixed, Travis got an error on jrb19, but it's an HTTP error downloading cucumber gem. Is it possible to rerun the suite on Travis?

mpasternacki commented:
The "gearwheel" menu on an individual build's page has a "restart build" option - started it just now.

mpasternacki commented:
Re update itself – I'm not 100% sure whether we want to push only the current tag (Vendor#tag_name), or all tags belonging to the module including the history (all tags in the repository beginning with Vendor#tag_name_base).

I'd rather lean towards pushing all tags – when we store details in git notes and are able to parse and review it, it would be less surprising to the end user if all notes were synchronized with upstream repository.

Pushing the same tag twice doesn't seem to be much of a problem - Git server will just reply that the ref is "up to date". It will be able to spot some problems too (like two checkouts running 'vendor sync' independently, and having the same tag pointing to different commits).

sharnik committed:
Push all the tags!

sharnik committed:
Small fixes.

mpasternacki commented:
Great! Merging, thanks.
